gem/flop_count.py

```diff
@@ -0,0 +1,192 @@
+        return expression_flops(a, parameters)
+    else:
+        return 1 + sum(expression_flops(child, parameters)
+                       for child in expr.children)
+
+
+@flops.register(gem.Sum)
+@flops.register(gem.Division)
+@flops.register(gem.Comparison)
+@flops.register(gem.MathFunction)
+@flops.register(gem.MinValue)
+@flops.register(gem.MaxValue)
+def flops_oneplus(expr, parameters):
+    return 1 + sum(expression_flops(child, parameters)
+                   for child in expr.children)
+
+
+@flops.register(gem.Power)
+def flops_power(expr, parameters):
+    base, exponent = expr.children
+    base_flops = expression_flops(base, parameters)
+    if isinstance(exponent, gem.Literal):
+        exponent = exponent.value
+        if exponent > 0 and exponent == math.floor(exponent):
+            return base_flops + int(math.ceil(math.log2(exponent)))
+    else:
+        return base_flops + 5   # heuristic
+
+
+@flops.register(gem.Conditional)
+def flops_conditional(expr, parameters):
+    condition, then, else_ = (expression_flops(child, parameters)
+                              for child in expr.children)
+    return condition + max(then, else_)
+
+
+@flops.register(gem.Indexed)
+@flops.register(gem.FlexiblyIndexed)
+def flops_indexed(expr, parameters):
+    aggregate = sum(expression_flops(child, parameters)
+                    for child in expr.children)
+    # Average flops per entry
+    return aggregate / numpy.product(expr.children[0].shape, dtype=int)
+
+
+@flops.register(gem.IndexSum)
+def flops_indexsum(expr, parameters):
+    raise ValueError("Not expecting IndexSum")
+
+
+@flops.register(gem.Inverse)
+def flops_inverse(expr, parameters):
+    n, _ = expr.shape
+    # 2n^3 + child flop count
+    return 2*n**3 + sum(expression_flops(child, parameters)
+                        for child in expr.children)
+
+
+@flops.register(gem.Solve)
+def flops_solve(expr, parameters):
+    n, m = expr.shape
+    # 2mn + inversion cost of A + children flop count
+    return 2*n*m + 2*n**3 + sum(expression_flops(child, parameters)
+                                for child in expr.children)
+
+
+@flops.register(gem.ComponentTensor)
+def flops_componenttensor(expr, parameters):
+    raise ValueError("Not expecting ComponentTensor")
+
+
+def expression_flops(expression, parameters, top=False):
+    """An approximation to flops required for each expression.
+
+    :arg expression: GEM expression.
+    :arg parameters: Useful miscellaneous information.
+    :arg top: are we at the root?
+    :returns: flop count for the expression
+    """
+    if not top and expression in parameters.temporaries:
+        return 0
+    else:
+        return flops(expression, parameters)
+
+
+def count_flops(impero_c):
+    """An approximation to flops required for a scheduled impero_c tree.
+
+    :arg impero_c: a :class:`~.Impero_C` object.
+    :returns: approximate flop count for the tree.
+    """
+    return statement(impero_c.tree, impero_c)
```

Comment:
I think not being able to count flops should be a soft failure, so can you wrap this in a try/except block catching the possible exceptions it might raise? return zero in the except.